### DISCUSSION TITLE
Help with custom realtime ScenarIO backend

### DISCUSSION THREAD

**Discussion by HorvathDawson** [robotology/gym-ignition#404] [Q&A]
I am starting to work on a scenario realtime back-end for my custom robot. Before I drive to deep into it with minimal cpp knowledge is there any knowledge transfer that would be useful for me? 

I only need it to support force controller mode. I already have my hardware control loop that does PID control (SO I think set_force_target should translate one to one for my force target on hardware? assuming PID constants are matched?). 

How would you recommend I format the project? Should I fork gym-ignition/scenarIO and work off of that? How would I make this integrate well into gym-ignition? 

Thank you for any advice :)
Dawson

Also thank you so much for the amazing work you ave done on this project !!

**Comment by diegoferigo** [comment]
In general you should be able to do everything from downstream, no need to operate directly upstream. Here below a possible task breakdown:

1. Develop a new real-time runtime that calls the environment at a desired rate and enforces the agent period to be followed. You can start from the placeholder [`realtime_runtime.py`](https://github.com/robotology/gym-ignition/blob/0a350b99c41c12ffe0f1eeda8a96046424ee30bd/python/gym_ignition/runtimes/realtime_runtime.py), copy it in your project, and play a bit with `time.sleep()` and similar.
1. Pick a language to implement the new ScenarIO backend:
   - If you can control your robot from Python, you can create new classes inheriting from `scenario.bindings.core.World|Model|Joint|Link` that call your Python module to get and set real robot data.
   - If you can control your robot from C++, you can create new classes inheriting from `scenario::core::World|Model|Joint|Link` that call your C++ APIs to get and set real robot data. Then, you should also create the bindings to expose this backend to Python, you can copy [what we do upstream with CMake](https://github.com/robotology/gym-ignition/tree/0a350b99c41c12ffe0f1eeda8a96046424ee30bd/scenario/bindings/gazebo) for the Gazebo backend.

Note that, while the ScenarIO APIs are quite a lot, you can implement only the methods that you need, leaving the others either as dummy or not implemented.

Once your backend is ready, if your task is developed in such a way that has no `*.to_gazebo` calls, you can replace the entrypoint (e.g. [this one](https://github.com/robotology/gym-ignition/blob/0a350b99c41c12ffe0f1eeda8a96046424ee30bd/python/gym_ignition_environments/__init__.py#L14) for the demo environments) specifying the real-time runtime instead of the gazebo runtime. The task should work also on the new real-time backend without changing any logic.

If you have randomizers that work for the gazebo back-end, most probably you have to develop new ones of the real-time back-end.

**Reply by HorvathDawson** [reply]
@diegoferigo  Thank you for the thorough answer. I have a few followup questions:

1. What is the purpose of enforcing an agent period inside the realtime_runtime? instead wouldn't it be better to handle that in the loop using the run-time class? (The real robot is an under actuated balancing task so real-time will be critical).
2. If I use C++ for the backend. When I create the new class inheriting from `scenario::core::World|Model|Joint|Link` how would the following work?
- If I used a similar Cmake structure as the gazebo backend how would it expose my realtime backend to python? Would it be imported similar to gazebo backend despite being downstream?
- Would it be necessary to install gym-ignition/scenarIO from source to do this downstream? Ideally I would like this package to be easy to install for my team after being developed. 
3. How do you recommend handling resets in real with the runtime?

**Reply by diegoferigo** [reply]
> @diegoferigo Thank you for the thorough answer. I have a few followup questions:
> 
>  1. What is the purpose of enforcing an agent period inside the realtime_runtime? instead wouldn't it be better to handle that in the loop using the run-time class? (The real robot is an under actuated balancing task so real-time will be critical).

The abstraction have been thought that way, the caller of `gym.Env.step` should not know anything about the need of stepping the environment, this means someone else inside the `gym.Env` object has to do it. The best candidate is the Runtime since in the real-time case does not have much to do beyond waiting the right moment to call the Task methods (as opposed to the Gazebo runtime that has to instantiate and step the simulator).

I would suggest to start with a runtime developed in Python, check what's the maximum rate that it can reach, and only then, if needed, try to optimize it with C++. In my experience, Python might suffice unless you want to reach kHz. What can be directly done in C++ is the ScenarIO backend, but this depends on the available library that you will use to interface with your robot (if it is in either C++ or Python).

> 2. If I use C++ for the backend. When I create the new class inheriting from `scenario::core::World|Model|Joint|Link` how would the following work?
>
>   * If I used a similar Cmake structure as the gazebo backend how would it expose my realtime backend to python? Would it be imported similar to gazebo backend despite being downstream?

I never tried to do it from downstream and something could be missing. I guess that you can create your own C++ backend by importing the CMake targets with `find_package(Scenario COMPONENTS Core REQUIRED)` (you can find a minimum example [here](https://github.com/ami-iit/gazebo-scenario-plugins/blob/f9d865ef4398e5c15dca589691762ca636d48b33/plugins/low_pass_target/CMakeLists.txt)). From there, you can create your own `custom::World|Model|Link|Joint` classes, and create the corresponding SWIG / pybind11 bindings[^1] (the learning curve could take some time for this task).

> * Would it be necessary to install gym-ignition/scenarIO from source to do this downstream? Ideally I would like this package to be easy to install for my team after being developed.

You can either install ScenarIO in developer mode, or install the `scenario` Python package and add its folder (`.../site-packages/scenario`) to `CMAKE_PREFIX_PATH` environment variable.

> 1. How do you recommend handling resets in real with the runtime?

Up to downstream, it varies for the task you're simulating. Maybe putting the robot in idle and provide the user some time, waiting a y/n feedback, for a manual reset? Just an idea.

[^1]: In the SWIG case, you can check [what we do for Gazebo](https://github.com/robotology/gym-ignition/blob/master/scenario/bindings/gazebo/gazebo.i). You need to import the swig file for core, not really sure how to do it properly by sourcing upstream.

**Reply by HorvathDawson** [reply]
@diegoferigo  Thank you so much for the help. I didnt want to waste any more of your time until I spent more time with the software. The learning curve for this has been pretty large and your help has made it slightly less daunting.

I have a few more small questions which I would really appreciate having a bit of help with:

I started by copying most of the structure that scenarI/O has. I do not need to implement anything from the link class so I ignored that class altogether. My initial goal is to make this backend build and compile as cpp before considering bindings. 

I used a lot of the CMake as a template for where I should begin with this. One thing I am a bit confused on is how I would add ScenarioCore CMake target as a dependency in my "install_basic_package_files"  [how I have it now](https://github.com/Baesian-Balancer/SIM/blob/d2f5ebe3e770cfff0d74dec59099dd1140853a0a/src/monopod/CMakeLists.txt#L44-L54). Would I need to o something like [this](https://github.com/Baesian-Balancer/SIM/blob/d2f5ebe3e770cfff0d74dec59099dd1140853a0a/src/monopod/CMakeLists.txt#L7) to import target? 

Is there a better way to set the CMAKE_PREFIX_PATH to ../site-packages/scenariol?
`set(ENV{CMAKE_PREFIX_PATH} ~/.local/lib/python3.8/site-packages/scenario)`

when you were developing the scenario packages what was the best method for building and testing code? Was there any good strategy to compile and test? Are there any blaring issues with what I have so far?

If you think it would be beneficial for this project I wouldn't be opposed to creating a short tutorial on how to make a scenarIO cpp backend downstream and set it up to link the bindings with the upstream bindings.

**Reply by diegoferigo** [reply]
> I started by copying most of the structure that scenarI/O has. I do not need to implement anything from the link class so I ignored that class altogether. My initial goal is to make this backend build and compile as cpp before considering bindings.

> when you were developing the scenario packages what was the best method for building and testing code? Was there any good strategy to compile and test? Are there any blaring issues with what I have so far?

This makes sense, I would recommend complementing the ScenarIO classes with a simple `main.cpp` so that you can test them step by step without the need to create Python bindings from the beginning. The repo does not currently have a cpp example since we're using Python as main target, but it should be quite straightforward writing one from scratch.

This approach forces you to think of a proper entrypoint from where to get all the ScenarIO objects. You have two choices:

1. Create something similar to `GazeboSimulator`. The main aim of this class is initialize all the resources and return a `World` object. Then, once you have the `World` object, you can use all the ScenarIO APIs as you like.
2. Initialize all the resources directly from the `World` class. This means that the `World` becomes your entrypoint, and in your `main.cpp`, the first thing you want to do is configuring and constructing this object with all the resources you need for interfacing with your robot. A while ago, I tried to develop a real-time backend for the YARP middleware (you can think of it as a ROS alternative) in [`feature/yarp`](https://github.com/diegoferigo/gym-ignition/tree/feature/yarp). It was a prototype that I never finished, but it was partially working. As you can see [here](https://github.com/diegoferigo/gym-ignition/blob/b83063bbb9ed85f28d92bb0193aa0563d792c9a5/cpp/scenario/yarp/src/World.cpp#L52-L55), I was passing to this class all the conf files needed to create a `Model` from which afterwards I could communicate with the real robot. I don't want to provide too many details since YARP usage is not the main point here, but if you check [`Model::jointPosition`](https://github.com/diegoferigo/gym-ignition/blob/b83063bbb9ed85f28d92bb0193aa0563d792c9a5/cpp/scenario/yarp/src/Model.cpp#L499-L538), you can see that I was gathering data from the YARP interfaces which read from the network (you can think of a ROS topic if you are familiar with ROS).

> I used a lot of the CMake as a template for where I should begin with this. One thing I am a bit confused on is how I would add ScenarioCore CMake target as a dependency in my "install_basic_package_files" [how I have it now](https://github.com/Baesian-Balancer/SIM/blob/d2f5ebe3e770cfff0d74dec59099dd1140853a0a/src/monopod/CMakeLists.txt#L44-L54).

The `install_basic_package_files` is a minor detail that you can initially skip. It is necessary only to distribute you code to downstream, and it is not necessary unless you want other people to use your ScenarIO backend. I would suggest to implement a proper way to distribute your project in a second moment.

> Would I need to o something like [this](https://github.com/Baesian-Balancer/SIM/blob/d2f5ebe3e770cfff0d74dec59099dd1140853a0a/src/monopod/CMakeLists.txt#L7) to import target?

> Is there a better way to set the CMAKE_PREFIX_PATH to ../site-packages/scenariol? `set(ENV{CMAKE_PREFIX_PATH} ~/.local/lib/python3.8/site-packages/scenario)`

This is the result of a proper`install_basic_package_files` configuration of ScenarIO Core. As written above, you don't need do to it unless you want to distribute your C++ code.

Reading these line, however, I noticed that you import the ScenarIO targets from the Python folder. This is ok, although if you are a developer that use C++ APIs, I would warmly recommend to switch to the Developer installation, that installs the entire C++ project outside the Python folder. Since Python is not an initial concern, it would provide a cleaner setup. In a Developer installation, if you install in the default system folder (you need root permissions), CMake is able to find the targets without the need to specify `CMAKE_PREFIX_PATH`.

> when you were developing the scenario packages what was the best method for building and testing code? Was there any good strategy to compile and test? Are there any blaring issues with what I have so far?

As I mentioned above, starting with a `main.cpp` is a good way to start with a low complexity.

> If you think it would be beneficial for this project I wouldn't be opposed to creating a short tutorial on how to make a scenarIO cpp backend downstream and set it up to link the bindings with the upstream bindings.

This is a good idea, feel free to open a new discussion in the future to provide some pointer for other users.

---

Quick note: if you remain in the C++ domain, I think everything is already in place here upstream. Instead, for extending the new backend with Python bindings, I suspect we need to start distributing somehow the `core.i` SWIG file. I have to think how to do it properly, initially it can be copied and pasted in downstream projects if needed.

**Reply by HorvathDawson** [reply]
Reply to old thread
-------------------------
> This approach forces you to think of a proper entrypoint from where to get all the ScenarIO objects. You have two choices:

I was planning (am currently) to use the world class as my entry point. Is there any reason I would not want to do that? How would using the world as a entry point work with the swig bindings? Would there be issues with that? 

> As you can see here, I was passing to this class all the conf files needed to create a Model from which afterwards I could communicate with the real robot.

So far the way I have been building the backend is much less general. Since I only plan to interface with a specific robot that wont be changing I haven't put any effort into using a config file etc to define the robot. Instead the Implementation details will be exposed from our Robot SDK. 

> This is the result of a properinstall_basic_package_files configuration of ScenarIO Core. As written above, you don't need do to it unless you want to distribute your C++ code. 

When you say distribute do you mean with pypi? I was hoping to eventually build our package as a pip installable package. However, that is not a concern at the moment so I will ignore this for now and take a look at it again in the future. Makes the CMake much simpler which I enjoy haha,

> Reading these line, however, I noticed that you import the ScenarIO targets from the Python folder. This is ok, although if you are a developer that use C++ APIs, I would warmly recommend to switch to the Developer installation, that installs the entire C++ project outside the Python folder. Since Python is not an initial concern, it would provide a cleaner setup. In a Developer installation, if you install in the default system folder (you need root permissions), CMake is able to find the targets without the need to specify CMAKE_PREFIX_PATH

I have a few questions about this. I have been using the python install for two reasons; Firstly I hope to make this package work as its own pip install that integrates into the python install of ignition gym. Second, I tried to install the developer version of ignition gym and ran into a bunch of issues especially with the IDynTree package. I spent some time trying to debug this but ultimately didn't want to spend to long forcing it to work. I could try to set it up again on a new user. 

> Quick note: if you remain in the C++ domain, I think everything is already in place here upstream. Instead, for extending the new backend with Python bindings, I suspect we need to start distributing somehow the core.i SWIG file. I have to think how to do it properly, initially it can be copied and pasted in downstream projects if needed.

Yes I have noticed that I will need the `core.i` file. Currently I have it copied into my downstream repo. It would be nice to have it exposed upstream though. 

Questions on Bindings
---------------------------
I am at the point where I want to start trying to install my backend as a plugin to the existing bindings.  I am slightly unsure what I need to change from the original CMake code you wrote for the original scenarIO project bindings. 

1. Would I be able to to replace all of [this](https://github.com/robotology/gym-ignition/blob/master/scenario/bindings/CMakeLists.txt#L58-L77) With  `add_subdirectory(my_subpackage)`?
2. Is there anything special I would have to do for CMake to find ScenarioSwig [here](https://github.com/robotology/gym-ignition/blob/master/scenario/bindings/gazebo/CMakeLists.txt#L23) when I replace gazebo with my downstream package?
3. Is there anything I would need to do for this to work with a virtual environment? I would rather not have to test installing the bindings on my main environment. Even more general did you find an easy way to test installing bindings? 
4. Is there anything other than the `.i` interface file and its corresponding CMake files that you think I would need to make to build the bindings? It doesn't seem like I need to make any `.init` file or other file for the bindings.

**Reply by HorvathDawson** [reply]
I am getting a error with the including of `core.i`. This is what I have currenty,
![image](https://user-images.githubusercontent.com/37032993/139964507-49dddfc5-b231-459a-9e7a-c4b4c50c09d9.png)

It doesnt seem like you do anything special to have swig find the components in the build tree so I am not sure what the issue is. My code is in this [branch](https://github.com/Baesian-Balancer/SIM/tree/dawson/dev) if you do want to take a look.

**Reply by HorvathDawson** [reply]
I was able to get swig to find the headers by linking the libraries like this,

```
target_link_libraries(${scenario_swig_name} PUBLIC ScenarioCore)

target_link_libraries(${scenario_swig_name} PUBLIC ScenarioMonopod)
```

However now I am getting some weird static cast issue with the class pointers,

> invalid static_cast from type ‘SwigValueWrapper<std::vector<std::shared_ptr<scenario::monopod::Model>, std::allocator<std::shared_ptr<scenario::monopod::Model> > > >’ to type ‘const std::vector<std::shared_ptr<scenario::core::Model> >&’
 
> no match for ‘operator=’ (operand types are ‘SwigValueWrapper<std::vector<std::shared_ptr<scenario::monopod::Model>, std::allocator<std::shared_ptr<scenario::monopod::Model> > > >’ and ‘std::vector<std::shared_ptr<scenario::core::Model> >’)

> no type named ‘type’ in ‘struct std::enable_if<false, void>’

**Reply by diegoferigo** [reply]
> ## Reply to old thread
> 
> > This approach forces you to think of a proper entrypoint from where to get all the ScenarIO objects. You have two choices:
> 
> I was planning (am currently) to use the world class as my entry point. Is there any reason I would not want to do that? How would using the world as a entry point work with the swig bindings? Would there be issues with that?

This looks good, I just wanted to point out two different alternatives. In the Gazebo ScenarIO backend, all the World / Model / Link / Classes are stateless, meaning that different objects created from the same simulation behave in the same way. This approach requires another class (`GazeboSimulator`) used to initialize these objects. I don't think this is a necessary feature in your case, I think you can proceed with World as entrypoint.

> > As you can see here, I was passing to this class all the conf files needed to create a Model from which afterwards I could communicate with the real robot.
> 
> So far the way I have been building the backend is much less general. Since I only plan to interface with a specific robot that wont be changing I haven't put any effort into using a config file etc to define the robot. Instead the Implementation details will be exposed from our Robot SDK.

Ok, this is up to you. Also on real robot, often a conf file is a good idea. Often there are multiple parameters accepted that otherwise should have to be hardcoded in the sources, but it could not be your case.

> > Reading these line, however, I noticed that you import the ScenarIO targets from the Python folder. This is ok, although if you are a developer that use C++ APIs, I would warmly recommend to switch to the Developer installation, that installs the entire C++ project outside the Python folder. Since Python is not an initial concern, it would provide a cleaner setup. In a Developer installation, if you install in the default system folder (you need root permissions), CMake is able to find the targets without the need to specify CMAKE_PREFIX_PATH
> 
> I have a few questions about this. I have been using the python install for two reasons; Firstly I hope to make this package work as its own pip install that integrates into the python install of ignition gym. Second, I tried to install the developer version of ignition gym and ran into a bunch of issues especially with the IDynTree package. I spent some time trying to debug this but ultimately didn't want to spend to long forcing it to work. I could try to set it up again on a new user.

The two setups are mostly the same (User and Developer). What's different is only where the headers and libraries of ScenarIO are installed. In order to let CMake find them, if they are not installed in one of the default folders, you have to configure manually your environment exporting `CMAKE_PREFIX_PATH`. You do this in your CMake, when instead it should be done by hand outside since the location could vary from system to system.


> I was able to get swig to find the headers by linking the libraries like this,
> 
> ```
> target_link_libraries(${scenario_swig_name} PUBLIC ScenarioCore)
> 
> target_link_libraries(${scenario_swig_name} PUBLIC ScenarioMonopod)
> ```

Yes, you need to link `ScenarioCore` against your bindings, it will add to your compiler command line the include path of the missing headers and link against the component libraries.

> However now I am getting some weird static cast issue with the class pointers,
> 
> > invalid static_cast from type ‘SwigValueWrapper<std::vector<std::shared_ptrscenario::monopod::Model, std::allocator<std::shared_ptrscenario::monopod::Model > > >’ to type ‘const std::vector<std::shared_ptrscenario::core::Model >&’
> 
> > no match for ‘operator=’ (operand types are ‘SwigValueWrapper<std::vector<std::shared_ptrscenario::monopod::Model, std::allocator<std::shared_ptrscenario::monopod::Model > > >’ and ‘std::vector<std::shared_ptrscenario::core::Model >’)
> 
> > no type named ‘type’ in ‘struct std::enable_if<false, void>’

Oh wow, SWIG problems are always funny -.- I have no idea what's going on here, at least for what concern the `SwigValueWrapper`. For the `std::enable_if`, are you sure you are not using this in a public header (included in your `.i` file)? Swig does not support modern C++ features.

**Reply by HorvathDawson** [reply]
I finally fixed the shared_ptr bug after way more hours than I would like to admit. 

I now have bindings for my cpp backend that compile downstream :) 

I plan to package these bindings as site-lib python package. I am a little unsure how I should do this. The issue is that it seems the only way to do this is to modify the existing scenario site-lib package to include my backend. this means I would have to install my bindings files, cpp headers, cmake files for my backend, and update the parent `ScenarioConfig.cmake` file to include my backends target.  I would also have to update the __init__ to import my swig bindings. 

I tested this by manually moving my files and it seemed to work. However, I am not sure how I would do this all in cmake. Also this does not feel like it would be very good practice modifying the existing Scenario package in the sitelib directly like this. 

I thought about making a "scenario_rt" package that is installed. The issue is I would have to modify my task classes to instead import this package. This is a big issue since I dont want the realtime backend to be coupled to the environments for cloud training. 

I would like this to be as simple as possible to install. Do you have any ideas?
